File: owntracks_to_db.py
#! /usr/bin/env python
"""
Store owntracks location updates in a database
"""
from datetime import datetime, timezone
from prometheus_client import start_http_server, Counter, Gauge
import json
import logging
import logging.handlers
import paho.mqtt.client as mqtt
import psycopg2
import ssl
import sys
import time
import yaml


class OwntracksToDatabaseBridge():
    def __init__(self, config='./.owntrackstodb.yaml'):
        try:
            with open(config, 'r') as f:
                self.total_recieved_updates = Counter(
                    'total_received_updates',
                    'The number of updates received from the mqtt broker')
                self.total_persisted_updates = Counter(
                    'total_persisted_updates',
                    'The number of updates saved into the database')
                configs = yaml.load(f.read())
                self._last_persisted_timestamp = 0
                self._last_received_timestamp = 0
                self.persistance_lag = Gauge(
                    'persistence_lag',
                    'Seconds between most-recently-received update and ' +
                    'the most-recently-persisted update')

                logging.basicConfig(level=logging.INFO)
                self._logger = logging.getLogger(__name__)
                formatter = logging.Formatter(
                    '%(asctime)s - %(name)s - %(levelname)s - %(message)s')
                handler = logging.handlers.RotatingFileHandler('o2db.log',
                                                               maxBytes=1048576,
                                                               backupCount=5)
                handler.setFormatter(formatter)
                self._logger.addHandler(handler)

                dbhost = configs['database']['host']
                dbport = configs['database']['port']
                dbuser = configs['database']['username']
                dbpass = configs['database']['password']
                dbname = configs['database']['dbname']

                self._conn = psycopg2.connect(host=dbhost, port=dbport,
                                              user=dbuser, password=dbpass,
                                              dbname=dbname)

                # Handle mqtt messages from the channels we subscribe to
                def handle_message(client, userdata, message):
                    j = json.loads(str(message.payload, encoding="ascii"))
                    if(message.topic.startswith("owntracks")):
                        # If message format is 'owntracks/user/device'
                        # then we should try to parse out userid/deviceid
                        # and handle this message as a location update
                        if(j['_type'] == 'location'):
                            userid = message.topic.split('/')[1]
                            device = message.topic.split('/')[2]
                            self._logger.info("{0} {1} posted an update: {2}"
                                              .format(userid, device, j))
                            self.total_recieved_updates.inc()
                            self.handle_location_update(userid, device, j)

                self._client = mqtt.Client(client_id="")
                try:
                    self._client.tls_set(ca_certs=configs['mqtt']['ca'],
                                         cert_reqs=ssl.CERT_REQUIRED,
                                         tls_version=ssl.PROTOCOL_TLSv1)
                except IOError as e:
                    self._logger.error("Something went wrong in mqtt setup. {0}"
                                       .format(e))
                self._client.username_pw_set(configs['mqtt']['username'],
                                             configs['mqtt']['password'])
                self._client.will_set("/lwt/o2db",
                                      payload="o2db script offline")
                mqtt_host = configs['mqtt']['host']
                mqtt_port = configs['mqtt']['port']
                self._client.on_message = handle_message
                self._logger.warn("Connecting to mqtt broker...")
                self._client.connect(mqtt_host, mqtt_port)
        except IOError as e:
            self._logger.error("Unable to load configuration. {0}".format(e))
            sys.exit(1)

    def handle_location_update(self, user, device, rawdata):
        acc = rawdata.get('acc')  # accuracy in meters
        alt = rawdata.get('alt')  # altitude above sea level
        batt = rawdata.get('batt')  # battery percentage
        cog = rawdata.get('cog')  # course over ground
        lat = rawdata['lat']  # latitude
        lon = rawdata['lon']  # longitude
        rad = rawdata.get('rad')  # radius around region
        t = rawdata.get('t')  # p, c, b, r, u, t; see booklet
        tid = rawdata.get('tid')  # tracker ID
        tst = datetime.fromtimestamp(rawdata['tst'], timezone.utc)  # Timestamp
        vac = rawdata.get('vac')  # vertical accuracy
        vel = rawdata.get('vel')  # velocity, kmh
        pressure = rawdata.get('p')  # barometric pressure in kPa (float)
        connection_status = rawdata.get('conn')  # connection status: w, o, m

        # execute prepared statement
        try:
            cur = self._conn.cursor()
            cur.execute(
                """insert into locationupdates (acc, alt, batt, cog, lat, lon,
                radius, t, tid, tst, vac, vel, p, conn, rawdata, userid, device)
                values (%(acc)s, %(alt)s, %(batt)s, %(cog)s, %(lat)s, %(lon)s,
                %(rad)s, %(t)s, %(tid)s, %(tst)s, %(vac)s, %(vel)s, %(p)s,
                %(conn)s, %(rawdata)s, %(userid)s, %(device)s);""",
                {'acc': acc, 'alt': alt, 'batt': batt, 'cog': cog, 'lat': lat,
                 'lon': lon, 'rad': rad, 't': t, 'tid': tid, 'tst': tst,
                 'vac': vac, 'vel': vel, 'p': pressure,
                 'conn': connection_status, 'rawdata': json.dumps(rawdata),
                 'userid': user, 'device': device})
            self._conn.commit()
            self.total_persisted_updates.inc()
        except Exception as e:
            # TODO We should try to persist this update again
            self._logger.error("Unable to execute query: {0}".format(e))

    def run(self):
        self._client.subscribe([("owntracks/#", 0)])
        self._logger.warn("subscribed to 'owntracks/#'")
        self._client.loop_forever()
        try:
            while True:
                time.sleep(2)
        except (KeyboardInterrupt, SystemExit):
            self._client.disconnect()

if __name__ == '__main__':
    start_http_server(8000)
    bridge = OwntracksToDatabaseBridge()
    bridge.run()

File: owntracks_to_db.rb
#!/opt/rh/ruby193/root/usr/bin/ruby
# -*- encoding: utf-8 -*-
#
# libraries
#   $ gem install mqtt
#   $ gem install pg
#   $ gem install logger
#

require 'bigdecimal'
require 'date'
require 'json'
require 'logger'
require 'mqtt'
require 'pg'
require 'rubygems'
require 'yaml'

class OwntracksUpdate
  # A struct to hold and validate an incoming message is a location update
  # before it is persisted to the database
  Version = '0.0.1'
  def initialize
  end
end


class OwntracksToDatabaseBridge
  # Subscribes to a mqtt broker for updates on the 'owntracks' channel,
  # validates the messages are location updates, and persists them to
  # a database
  Version = '0.0.1'

  def initialize(arguments, stdin)
    configuration = YAML::load_file(File.join(File.dirname(__FILE__), '.owntrackstodb.yaml'))
    @mqtt = MQTT::Client.new
    @mqtt.host = configuration['mqtt']['host'] 
    @mqtt.port = configuration['mqtt']['port']
    @mqtt.ssl = configuration['mqtt']['ssl']
    @mqtt.ca_file = configuration['mqtt']['ca_cert']
    @mqtt.username = configuration['mqtt']['username']
    @mqtt.password = configuration['mqtt']['password']

    # set up database configuration from yaml 
    # and connect to database
    dbhost = configuration['database']['host']
    dbport = configuration['database']['port']
    dbuser = configuration['database']['username']
    dbpass = configuration['database']['password']
    dbname = configuration['database']['dbname']
    @conn = PG.connect(:host => dbhost, :port => dbport, :user => dbuser, :password => dbpass, :dbname => dbname)
    @conn.prepare('locationupdate1', 'insert into locationupdates (acc, alt, batt, cog, lat, lon, radius, t, tid, tst, vac, vel, p, conn, rawdata, userid, device) values ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, $16, $17)')

    STDOUT.sync = true
    #@log = Logger.new(STDOUT)
    @log = Logger.new('owntrackstodb.log', 10, 1048576)
    @log.level = Logger::INFO

  end

  def handle_location_update(user, device, json)
    acc = json['acc'] # accuracy
    alt = json['alt'] # altitude
    batt = json['batt'] # battery percentage
    cog = json['cog'] # course over ground
    lat = json['lat'] # latitude
    lon = json['lon'] # longitude
    rad = json['rad'] # radius around region
    t  = json['t'] # p, c, b, r, u, t; see booklet
    tid = json['tid'] # tracker ID
    tst = Time.at(json['tst']).to_datetime # timestamp in epoch format
    vac = json['vac'] # vertical accuracy
    vel = json['vel'] # velocity, kmh
    pressure = json['p'] # barometric pressure in kPa (float)
    connection_status= json['conn'] # connection status: w, o, m; see booklet
    rawdata = json
    @log.info("rawdata is #{json}")

    # now that we have all the elements we need, we should insert them into the database
    @conn.exec_prepared('locationupdate1', [acc, alt, batt, cog, lat, lon, rad, t, tid, tst, vac, vel, pressure, connection_status, rawdata, user, device])

  end

  def handle_message(topic, message)
    @log.info("received mqtt message : topic=#{topic}, message=#{message}")
    # Handle a single owntracks message
    # topic will be 'owntracks/user/device'
    # the message will be a json hash
    user, device = topic.split("/")[1], topic.split("/")[2]
    json = JSON.parse(message)
    @log.info("#{user} #{device} posted an update: #{json}")
    # example message
    #  {"tst"=>1493917547, "acc"=>65, "_type"=>"location", "alt"=>19, "lon"=>-122.6809371528689, "vac"=>21, "p"=>101.0912551879883, "lat"=>45.52250024956788, "batt"=>52, "conn"=>"m", "tid"=>"jv"}
    if (json['_type'] == 'location')
      handle_location_update(user, device, json)
    end
  end

  def run
    begin
      @mqtt.connect() do |c|
        @log.info("Connection start...")
        c.get('owntracks/#') do |topic, message|
          handle_message topic, message
        end 
      end # end of connection
    rescue Exception => e
      @log.error(e)
    end
  end #end of run block
end #end class def

app = OwntracksToDatabaseBridge.new(ARGV, STDIN)
app.run
